optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py

```diff
@@ -0,0 +1,331 @@
+    """
+
+    name = 'cit'
+    description = "CIT Varianter options for the 'run' subcommand"
+
+    def configure(self, parser):
+
+        for name in ("run", "variants"):
+            subparser = parser.subcommands.choices.get(name, None)
+            if subparser is None:
+                continue
+            cit = subparser.add_argument_group('CIT varianter options')
+            cit.add_argument('--cit-parameter-file', metavar='PATH',
+                             help="Paths to a parameter file")
+            cit.add_argument('--cit-parameter-path', metavar='PATH',
+                             default='/run',
+                             help=('Default path for parameters generated '
+                                   'on the CIT variants'))
+            cit.add_argument('--cit-order-of-combinations',
+                             metavar='ORDER', type=int, default=2,
+                             help=("Order of combinations. Defaults to "
+                                   "%(default)s, maximum number is specific "
+                                   "to parameter file content"))
+
+    def run(self, args):
+        pass
+
+
+class VarianterCit(Varianter):
+
+    """
+    Processes the parameters file into variants
+    """
+
+    name = 'cit'
+    description = "CIT Varianter"
+
+    def initialize(self, args):
+        self.variants = None
+
+        cit_parameter_file = getattr(args, "cit_parameter_file", None)
+        if cit_parameter_file is None:
+            return
+        else:
+            cit_parameter_file = os.path.expanduser(cit_parameter_file)
+            if not os.access(cit_parameter_file, os.R_OK):
+                LOG_UI.error("parameter file '%s' could not be found or "
+                             "is not readable", cit_parameter_file)
+                self.error_exit(args)
+
+        self.parameter_path = getattr(args, "cit_parameter_path")
+        config = configparser.ConfigParser()
+        try:
+            config.read(cit_parameter_file)
+        except Exception as details:
+            LOG_UI.error("Cannot parse parameter file: %s", details)
+            self.error_exit(args)
+
+        parameters = [(key, value.split(', '))
+                      for key, value in config.items('parameters')]
+        constraints = [(key, value.split(', '))
+                       for key, value in config.items('constraints')]
+        order = args.cit_order_of_combinations
+        cit = Cit(parameters, constraints, order)
+        self.headers, self.variants = cit.combine()
+
+    @staticmethod
+    def error_exit(args):
+        if args.subcommand == 'run':
+            sys.exit(exit_codes.AVOCADO_JOB_FAIL)
+        else:
+            sys.exit(exit_codes.AVOCADO_FAIL)
+
+    def __iter__(self):
+        if self.variants is None:
+            return
+
+        variant_ids = []
+        for variant in self.variants:
+            base_id = "-".join([variant.get(key) for key in self.headers])
```

Comment:
This results in quite long names as you include full content of that parameter as part of base-id. Is this expected? For 6 dummy parameters it works well and gives great overview, but imagine real-world scenario with 100 params including full paths etc.:

```
avocado variants --cit-parameter-file examples/varianter_cit/params.ini
CIT Variants (8):
Variant /path/to/build/qemu-system-x86_64-virtio_scsi-http://ftp.fi.muni.cz/pub/linux/fedora/linux/releases/28/Server/x86_64/os/-512m-kvm-q35-3e2c:    /run
Variant /usr/libexec/qemu-system-x86_64-virtio_scsi-/path/to/iso-4g-tcg-i440fx-7643:    /run
Variant /usr/libexec/qemu-system-x86_64-virtio_blk-http://ftp.fi.muni.cz/pub/linux/fedora/linux/releases/28/Server/x86_64/os/-512m-kvm-i440fx-695c:    /run
Variant /path/to/build/qemu-system-x86_64-virtio_blk-/path/to/iso-4g-tcg-q35-9a73:    /run
Variant /path/to/build/qemu-system-x86_64-virtio_scsi-/path/to/iso-512m-kvm-i440fx-680d:    /run
Variant /usr/libexec/qemu-system-x86_64-virtio_scsi-http://ftp.fi.muni.cz/pub/linux/fedora/linux/releases/28/Server/x86_64/os/-4g-tcg-q35-0360:    /run
Variant /path/to/build/qemu-system-x86_64-virtio_blk-/path/to/iso-4g-kvm-i440fx-069f:    /run
Variant /usr/libexec/qemu-system-x86_64-virtio_scsi-/path/to/iso-512m-tcg-i440fx-8e23:    /run
```

Reply:
Btw that also stores all of the values in memory throughout the test execution...

Reply:
It's not really an Avocado problem, it allows the plugin to specify any name that is unique.

The problem with storing all values in memory is also not Avocado-vt problem, it only gets the values when executing each test. But this plugin generates all `base_ids` and each ID contains all of the values of that variant. For example those are values of "Avocado-vt" test:

```
root:     _name_map_file = {'subtests.cfg': '(subtest=io-github-autotest-qemu).(subtest=boot)', 'guest-hw.cfg': 'bridge.default_bios.no_virtio_rng.(image_backend=filesystem).no_9p_export.smallpages.no_pci_assignable.qcow2.virtio_scsi.smp2.virtio_net', 'host-os.cfg': 'Host.Fedora.m27.u0.Host_arch_x86_64', 'guest-os.cfg': 'Guest.Linux.JeOS.27.x86_64', 'machines.cfg': 'i440fx', 'tests.cfg': 'qemu_kvm_jeos_quick'}
root:     _short_name_map_file = {'subtests.cfg': 'io-github-autotest-qemu.boot', 'guest-hw.cfg': 'bridge.default_bios.no_virtio_rng.filesystem.no_9p_export.smallpages.no_pci_assignable.qcow2.virtio_scsi.smp2.virtio_net', 'host-os.cfg': 'Host.Fedora.m27.u0.Host_arch_x86_64', 'guest-os.cfg': 'Guest.Linux.JeOS.27.x86_64', 'machines.cfg': 'i440fx', 'tests.cfg': 'qemu_kvm_jeos_quick'}
root:     auto_cpu_model = yes
root:     backup_dir = images/
root:     backup_image = no
root:     backup_image_before_testing = yes
root:     backup_image_on_check_error = no
root:     boot_menu = off
root:     boot_once = c
root:     boot_order = cdn
root:     boot_path = images/pxeboot
root:     catch_monitor = catch_monitor
root:     cd_format = scsi-cd
root:     cdrom_check_cdrom_pattern = /dev/cdrom-\w+|/dev/cdrom\d*
root:     cdrom_get_cdrom_cmd = ls /dev/cdrom*
root:     cdrom_info_cmd = cat /proc/sys/dev/cdrom/info
root:     cdrom_test_cmd = dd if=%s of=/dev/null bs=1 count=1
root:     cdroms = cd1
root:     check_image = yes
root:     cmds_installed_host = 
root:     connect_uri = default
root:     convert_ppm_files_to_png = no
root:     cpu_chk_cmd = grep -c "^processor\b" /proc/cpuinfo
root:     cpu_model_flags = 
root:     create_vm_libvirt = no
root:     dep = ['qemu_kvm_jeos_quick.bridge.default_bios.no_virtio_rng.(image_backend=filesystem).no_9p_export.smallpages.no_pci_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Linux.JeOS.27.x86_64.Host.Fedora.m27.u0.Host_arch_x86_64.install', 'qemu_kvm_jeos_quick.bridge.default_bios.no_virtio_rng.(image_backend=filesystem).no_9p_export.smallpages.no_pci_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Linux.JeOS.27.x86_64.Host.Fedora.m27.u0.Host_arch_x86_64.setup', 'qemu_kvm_jeos_quick.bridge.default_bios.no_virtio_rng.(image_backend=filesystem).no_9p_export.smallpages.no_pci_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Linux.JeOS.27.x86_64.Host.Fedora.m27.u0.Host_arch_x86_64.image_copy', 'qemu_kvm_jeos_quick.bridge.default_bios.no_virtio_rng.(image_backend=filesystem).no_9p_export.smallpages.no_pci_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Linux.JeOS.27.x86_64.Host.Fedora.m27.u0.Host_arch_x86_64.unattended_install.cdrom']
root:     display = vnc
root:     download_url_stress = http://people.seas.harvard.edu/~apw/stress/stress-1.0.4.tar.gz
root:     drive_format = scsi-hd
root:     drive_index_cd1 = 1
root:     drive_index_image1 = 0
root:     drive_rerror_image1 = 
root:     drive_werror_image1 = 
root:     enable_guest_sosreport = no
root:     enable_host_sosreport = no
root:     enable_libvirtd_debug_log = yes
root:     enable_msix_vectors = yes
root:     enable_pvpanic = yes
root:     enable_remote_host_sosreport = no
root:     encode_video_files = yes
root:     env_cleanup = no
root:     file_transfer_client = scp
root:     file_transfer_port = 22
root:     flexible_nic_index = no
root:     guest_dmesg_ignore = False
root:     guest_dmesg_level = 3
root:     guest_port_remote_shell = 22
root:     id = io-github-autotest-qemu.boot
root:     image_backend = filesystem
root:     image_chain = 
root:     image_clone_command = cp --reflink=auto %s %s
root:     image_format = qcow2
root:     image_name = images/jeos-27-x86_64
root:     image_raw_device = no
root:     image_remove_command = rm -rf %s
root:     image_size = 10G
root:     image_unbootable_pattern = Hard Disk.*not a bootable disk
root:     image_verify_bootable = yes
root:     images = image1
root:     inactivity_treshold = 1800
root:     inactivity_watcher = log
root:     indirect_image_blacklist = /dev/hda[\d]* /dev/sda[\d]* /dev/sg0 /dev/md0
root:     install_virtio = no
root:     ioport_pvpanic = 0x505
root:     keep_ppm_files = no
root:     keep_ppm_files_on_error = no
root:     keep_screendumps = yes
root:     keep_screendumps_on_error = yes
root:     keep_video_files = yes
root:     keep_video_files_on_error = yes
root:     kill_timeout = 60
root:     kill_unresponsive_vms = yes
root:     kill_vm = yes
root:     kill_vm_before_test = no
root:     kill_vm_gracefully = yes
root:     kill_vm_libvirt = no
root:     kill_vm_on_error = yes
root:     libvirt_controller = virtio-scsi
root:     libvirtd_debug_file = 
root:     libvirtd_debug_level = 1
root:     local_ip = ENTER.YOUR.LOCAL.EXAMPLE.COM
root:     local_pwd = 
root:     login_timeout = 240
root:     machine_type = pc
root:     main_vm = avocado-vt-vm1
root:     malloc_perturb = yes
root:     mem = 1024
root:     mem_chk_cmd = dmidecode -t 17 | awk -F: '/Size/ {print $2}'
root:     mem_chk_cur_cmd = grep MemTotal /proc/meminfo
root:     mem_chk_re_str = [^\$]([0-9]+)
root:     memballoon_model = virtio
root:     migrate_dest_host = ENTER.YOUR.DEST.EXAMPLE.COM
root:     migrate_dest_pwd = PASSWORD.DEST.EXAMPLE
root:     migrate_load_vms = 
root:     migrate_main_vm = avocado-vt-vm1
root:     migrate_port = 49152
root:     migrate_proto = tcp
root:     migrate_shared_storage = SHARED_IMAGE.EXAMPLE
root:     migrate_source_host = ENTER.YOUR.SOURCE.EXAMPLE.COM
root:     migrate_source_pwd = PASSWORD.SOURCE.EXAMPLE
root:     migrate_vms = 
root:     migration_setup = no
root:     monitor_type = qmp
root:     monitors = qmpmonitor1
root:     name = qemu_kvm_jeos_quick.bridge.default_bios.no_virtio_rng.(image_backend=filesystem).no_9p_export.smallpages.no_pci_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Linux.JeOS.27.x86_64.Host.Fedora.m27.u0.Host_arch_x86_64.(subtest=io-github-autotest-qemu).(subtest=boot).i440fx
root:     netdev_peer_re = \s{2,}(.*?): .*?\\\s(.*?):
root:     netdst = virbr0
root:     nettype = user
root:     nic_model = virtio
root:     nics = nic1
root:     os_type = linux
root:     os_variant = fedora27
root:     other_tests_dirs = 
root:     password = 123456
root:     paused_after_start_vm = no
root:     pci_assignable = no
root:     profilers = kvm_stat
root:     provider = io-github-autotest-qemu
root:     qemu_binary = /bin/qemu-kvm
root:     qemu_dst_binary = /bin/qemu-kvm
root:     qemu_img_binary = /bin/qemu-img
root:     qemu_io_binary = /bin/qemu-io
root:     reboot_command = shutdown -r now
root:     redirs = remote_shell
root:     remote_ip = ENTER.YOUR.REMOTE.EXAMPLE.COM
root:     remote_pwd = 
root:     remote_user = root
root:     remove_image_image1 = no
root:     restart_vm = yes
root:     restore_image = no
root:     restore_image_after_testing = yes
root:     restore_image_on_check_error = no
root:     run_tcpdump = no
root:     screendump_delay = 5
root:     screendump_quality = 30
root:     screendump_temp_dir = /dev/shm
root:     screendump_verbose = no
root:     scsi_hba = virtio-scsi-pci
root:     serials = serial0
root:     shell_client = ssh
root:     shell_linesep = \n
root:     shell_port = 22
root:     shortname = qcow2.virtio_scsi.smp2.virtio_net.Guest.JeOS.27.x86_64.io-github-autotest-qemu.boot
root:     shutdown_command = shutdown -h now
root:     skip_cluster_leak_warn = no
root:     skip_image_check_during_running = no
root:     skip_image_processing = no
root:     smp = 2
root:     start_vm = yes
root:     status_test_command = echo $?
root:     store_vm_register = yes
root:     stress_args = --cpu 4 --io 4 --vm 2 --vm-bytes 256M
root:     strict_mode = no
root:     subtest = io-github-autotest-qemu.boot
root:     take_regular_screendumps = yes
root:     test_timeout = 14400
root:     type = boot
root:     url_auto_ip = 
root:     usb_bus = usb1.0
root:     usb_controller = ehci
root:     usb_devices = tablet1
root:     usb_max_port = 6
root:     usb_type = ich9-usb-ehci1
root:     usb_type_tablet1 = usb-tablet
root:     usb_type_usb1 = ich9-usb-ehci1
root:     usbs = usb1
root:     use_autostart = no
root:     use_check_cpu = yes
root:     use_debug = yes
root:     use_no_reboot = no
root:     use_os_type = yes
root:     use_os_variant = no
root:     use_virt_install_wait = no
root:     used_cpus = 2
root:     used_mem = 512
root:     username = root
root:     vcpu_thread_pattern = thread_id.?[:|=]\s*(\d+)
root:     verify_guest_dmesg = yes
root:     verify_host_dmesg = yes
root:     vfd_size = 1440k
root:     vga = std
root:     virt_install_wait_time = 300
root:     virt_test_type = qemu libvirt
root:     virtinstall_extra_args = 
root:     vm_arch_name = x86_64
root:     vm_register_delay = 5
root:     vm_type = qemu
root:     vms = avocado-vt-vm1
root:     vnc_autoport = yes
```

which would result in `{\'subtests.cfg\': \'(subtest=io-github-autotest-qemu).(subtest=boot)\', \'guest-hw.cfg\': \'bridge.default_bios.no_virtio_rng.(image_backend=filesystem).no_9p_export.smallpages.no_pci_assignable.qcow2.virtio_scsi.smp2.virtio_net\', \'host-os.cfg\': \'Host.Fedora.m27.u0.Host_arch_x86_64\', \'guest-os.cfg\': \'Guest.Linux.JeOS.27.x86_64\', \'machines.cfg\': \'i440fx\', \'tests.cfg\': \'qemu_kvm_jeos_quick\'}-{\'subtests.cfg\': \'io-github-autotest-qemu.boot\', \'guest-hw.cfg\': \'bridge.default_bios.no_virtio_rng.filesystem.no_9p_export.smallpages.no_pci_assignable.qcow2.virtio_scsi.smp2.virtio_net\', \'host-os.cfg\': \'Host.Fedora.m27.u0.Host_arch_x86_64\', \'guest-os.cfg\': \'Guest.Linux.JeOS.27.x86_64\', \'machines.cfg\': \'i440fx\', \'tests.cfg\': \'qemu_kvm_jeos_quick\'}-yes-images/-no-yes-no-off-c-cdn-images/pxeboot-catch_monitor-scsi-cd-/dev/cdrom-\\w+|/dev/cdrom\\d*-ls /dev/cdrom*-cat /proc/sys/dev/cdrom/info-dd if=%s of=/dev/null bs=1 count=1-cd1-yes--default-no-grep -c "^processor\x08" /proc/cpuinfo--no-[\'qemu_kvm_jeos_quick.bridge.default_bios.no_virtio_rng.(image_backend=filesystem).no_9p_export.smallpages.no_pci_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Linux.JeOS.27.x86_64.Host.Fedora.m27.u0.Host_arch_x86_64.install\', \'qemu_kvm_jeos_quick.bridge.default_bios.no_virtio_rng.(image_backend=filesystem).no_9p_export.smallpages.no_pci_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Linux.JeOS.27.x86_64.Host.Fedora.m27.u0.Host_arch_x86_64.setup\', \'qemu_kvm_jeos_quick.bridge.default_bios.no_virtio_rng.(image_backend=filesystem).no_9p_export.smallpages.no_pci_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Linux.JeOS.27.x86_64.Host.Fedora.m27.u0.Host_arch_x86_64.image_copy\', \'qemu_kvm_jeos_quick.bridge.default_bios.no_virtio_rng.(image_backend=filesystem).no_9p_export.smallpages.no_pci_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Linux.JeOS.27.x86_64.Host.Fedora.m27.u0.Host_arch_x86_64.unattended_install.cdrom\']-vnc-http://people.seas.harvard.edu/~apw/stress/stress-1.0.4.tar.gz-scsi-hd-1-0---no-no-yes-yes-yes-no-yes-no-scp-22-no-False-3-22-io-github-autotest-qemu.boot-filesystem--cp --reflink=auto %s %s-qcow2-images/jeos-27-x86_64-no-rm -rf %s-10G-Hard Disk.*not a bootable disk-yes-image1-1800-log-/dev/hda[\\d]* /dev/sda[\\d]* /dev/sg0 /dev/md0-no-0x505-no-no-yes-yes-yes-yes-60-yes-yes-no-yes-no-yes-virtio-scsi--1-ENTER.YOUR.LOCAL.EXAMPLE.COM--240-pc-avocado-vt-vm1-yes-1024-dmidecode -t 17 | awk -F: \'/Size/ {print $2}\'-grep MemTotal /proc/meminfo-[^\\$]([0-9]+)-virtio-ENTER.YOUR.DEST.EXAMPLE.COM-PASSWORD.DEST.EXAMPLE--avocado-vt-vm1-49152-tcp-SHARED_IMAGE.EXAMPLE-ENTER.YOUR.SOURCE.EXAMPLE.COM-PASSWORD.SOURCE.EXAMPLE--no-qmp-qmpmonitor1-qemu_kvm_jeos_quick.bridge.default_bios.no_virtio_rng.(image_backend=filesystem).no_9p_export.smallpages.no_pci_assignable.qcow2.virtio_scsi.smp2.virtio_net.Guest.Linux.JeOS.27.x86_64.Host.Fedora.m27.u0.Host_arch_x86_64.(subtest=io-github-autotest-qemu).(subtest=boot).i440fx-\\s{2,}(.*?): .*?\\\\s(.*?):-virbr0-user-virtio-nic1-linux-fedora27--123456-no-no-kvm_stat-io-github-autotest-qemu-/bin/qemu-kvm-/bin/qemu-kvm-/bin/qemu-img-/bin/qemu-io-shutdown -r now-remote_shell-ENTER.YOUR.REMOTE.EXAMPLE.COM--root-no-yes-no-yes-no-no-5-30-/dev/shm-no-virtio-scsi-pci-serial0-ssh--22-qcow2.virtio_scsi.smp2.virtio_net.Guest.JeOS.27.x86_64.io-github-autotest-qemu.boot-shutdown -h now-no-no-no-2-yes-echo $?-yes---cpu 4 --io 4 --vm 2 --vm-bytes 256M-no-io-github-autotest-qemu.boot-yes-14400-boot--usb1.0-ehci-tablet1-6-ich9-usb-ehci1-usb-tablet-ich9-usb-ehci1-usb1-no-yes-yes-no-yes-no-no-2-512-root-thread_id.?[:|=]\\s*(\\d+)-yes-yes-1440k-std-300-qemu libvirt--x86_64-5-qemu-avocado-vt-vm1-yes-1b3a` name and all the values being stored throughout the execution in memory (including all the values of the other variants).

Reply:
... so I'd recommend either serial-id or for configs with `< SOME_AMOUNT_OF_PARAMS` to provide indexes of the variants, therefor instead of:

```
q35-blk-smallpages-git-1bc2
q35-blk-hugepages-git-3d7e
```

it'd be:

```
1-1-1-1
1-1-2-1
```

and instead of:

```
foo-bar-baz-aaa-bbb-ccc-ddd-3d79
```

it'd be

```
variant1
```

(there is no need for random characters as each variant is guarantee to be used once)

Reply:
Well (luckily) there is no "standard" way of naming ones kids (eg. provided by state to those who chose to)... what I mean is that there is no really universal solution which means plugins have to chose based on their structure.

As for the Avocado-vt example, yes, it was to demonstrate "insane" example, but combinatorial is meant to simplify complex matrices so we might expect more than 10 axes, where some values might get quite long.

Anyway what do you think about the proposal? I'd suggest in `header` to print the axes mapping and in ids to use axes indexes only. Later (when needed) I'd also add the fallback to serial indexes for configs with > 20 (or so) axes.